cpp/include/cudf/io/memory_resource.hpp

```diff
@@ -41,4 +41,13 @@ rmm::host_async_resource_ref set_host_memory_resource(rmm::host_async_resource_r
  */
 rmm::host_async_resource_ref get_host_memory_resource();
 
+/**
+ * @brief Configure the size of the default host memory resource.
+ *
+ * Must be called before any other function in this header.
```

Comment:
This is a dangerous requirement. How about making the static function re-configuruable?

Reply:
For doing so:
 1. Static variable is declared outside of function scope (but in an anonymous namespace, so it is static inside just this TU). In addition, it can be a smart pointer.
 2. `host_mr` will initialize it with `std::nullopt` size if it is `nullptr`, otherwise just derefs the current pointer and returns.
 3. User can specify a size parameter to overwrite that static variable.
 4. All these ops should be thread safe.

Reply:
to clarify, the issue with calling config after get/set is that it would have no effect.

allowing this opens another can of worms., e.g. what is the intended effect of calling config after set?

Reply:
If we don't allow this, let's make some validity check to prevent it from being accidentally misused.

Reply:
@abellina what behavior do you suggest when config is called after the first resource use? I'm not sure if we should throw or just warn.

Reply:
I think we should throw, I agree with @ttnghia that we should do something in that case.

Reply:
Added a mechanism to throw if config is called after the default resource has already been created.
@abellina might be good to test your branch with this change.